File: android/src/main/kotlin/com/cloudwebrtc/webrtc/controller/RtpSenderController.kt
package com.cloudwebrtc.webrtc.controller

import com.cloudwebrtc.webrtc.TrackRepository
import com.cloudwebrtc.webrtc.proxy.MediaStreamTrackProxy
import com.cloudwebrtc.webrtc.proxy.RtpSenderProxy
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Controller for the [RtpSenderProxy] functional.
 *
 * @param messenger messenger used for creating new [MethodChannel]s.
 * @property track underlying [RtpSenderProxy] on which method calls will be performed.
 */
class RtpSenderController(messenger: BinaryMessenger, private val sender: RtpSenderProxy) :
        MethodChannel.MethodCallHandler, IdentifiableController {
    /**
     * Unique ID of the [MethodChannel] of this controller.
     */
    private val channelId = nextChannelId()

    /**
     * Channel which will be listened for the [MethodCall]s.
     */
    private val methodChannel =
            MethodChannel(messenger, ChannelNameGenerator.withId("RtpSender", channelId))

    init {
        methodChannel.setMethodCallHandler(this);
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setTrack" -> {
                val trackId: String? = call.argument("trackId")
                val track = if (trackId != null) {
                    TrackRepository.getTrack(trackId)!!
                } else {
                    null
                }
                sender.setTrack(track)
                result.success(null)
            }
            "dispose" -> {
                dispose()
                result.success(null)
            }
        }
    }

    /**
     * Converts this [RtpSenderController] to the Flutter's method call result.
     *
     * @return [Map] generated from this controller which can be returned to the Flutter side.
     */
    fun asFlutterResult(): Map<String, Any> = mapOf("channelId" to channelId)

    /**
     * Closes method channel of this [RtpSenderController].
     */
    private fun dispose() {
        methodChannel.setMethodCallHandler(null)
    }
}

File: ios/Classes/controller/RtpSenderController.swift
import Flutter

/// Controller for the `RtpSender`.
public class RtpSenderController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of `RtpSender` proxy.
  private var rtpSender: RtpSenderProxy

  /// ID of this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Creates new `RtpSenderController` for the provided `RtpSenderProxy`.
  init(messenger: FlutterBinaryMessenger, rtpSender: RtpSenderProxy) {
    let channelName = ChannelNameGenerator.name(name: "RtpSender", id: channelId)
    self.messenger = messenger
    self.rtpSender = rtpSender
    self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    self.channel.setMethodCallHandler({ (call, result) in
      try! self.onMethodCall(call: call, result: result)
    })
  }

  /// Handles all supported Flutter method calls for the `RtpSenderProxy`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) throws {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "replaceTrack":
      let trackId = argsMap!["trackId"] as? String
      var track: MediaStreamTrackProxy?
      if trackId != nil {
        track = MediaStreamTrackStore.tracks[trackId!]
      } else {
        track = nil
      }
      self.rtpSender.replaceTrack(t: track)
      result(nil)
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller to the Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    return [
      "channelId": self.channelId
    ]
  }
}
